File: 10 - Product Catlogue/product-catalogue-frontend/src/services/productService.js
import axios from "axios";

const API = "http://localhost:8080/api/products";

export const getProducts = () => axios.get(API);

export const getProduct = (id) =>
    axios.get(`${API}/${id}`);

export const addProduct = (product) =>
    axios.post(API, product);

export const updateProduct = (id, product) =>
    axios.put(`${API}/${id}`, product);

export const deleteProduct = (id) =>
    axios.delete(`${API}/${id}`);

// Filters

export const getByCategory = (category) =>
    axios.get(`${API}/category/${category}`);

export const getByBrand = (brand) =>
    axios.get(`${API}/brand/${brand}`);

export const getByAvailability = (available) =>
    axios.get(`${API}/availability/${available}`);

export const getByPrice = (min, max) =>
    axios.get(`${API}/price?min=${min}&max=${max}`);

// Filter by Category
export const filterByCategory = (category) =>
    axios.get(`${API}/category/${category}`);

// Filter by Brand
export const filterByBrand = (brand) =>
    axios.get(`${API}/brand/${brand}`);

// Filter by Availability
export const filterByAvailability = (available) =>
    axios.get(`${API}/availability/${available}`);

// Filter by Price
export const filterByPrice = (min, max) =>
    axios.get(`${API}/price?min=${min}&max=${max}`);

File: 10 - Product Catlogue/productcatlogue/src/main/java/com/example/productcatlogue/service/ProductService.java
package com.example.productcatlogue.service;

import com.example.productcatlogue.entity.Product;

import java.util.List;

public interface ProductService {

    Product addProduct(Product product);

    List<Product> getAllProducts();

    Product getProductById(Long id);

    Product updateProduct(Long id, Product product);

    void deleteProduct(Long id);

    List<Product> getByCategory(String category);

    List<Product> getByBrand(String brand);

    List<Product> getByAvailability(boolean available);

    List<Product> getByPriceRange(double min, double max);

    List<Product> getCategoryAndBrand(String category, String brand);

    List<Product> getCategoryAndAvailability(String category, boolean available);

    List<Product> getBrandAndAvailability(String brand, boolean available);

    List<Product> getCategoryAndPrice(String category, double min, double max);
}